CREATE TABLE association_by_datasource_direct(datatypeId TEXT, datasourceId TEXT, diseaseId TEXT, targetId TEXT, score FLOAT8, evidenceCount BIGINT);;

CREATE TABLE association_by_datasource_indirect(datatypeId TEXT, datasourceId TEXT, diseaseId TEXT, targetId TEXT, score FLOAT8, evidenceCount BIGINT);;

CREATE TABLE association_by_datatype_direct(diseaseId TEXT, targetId TEXT, datatypeId TEXT, score FLOAT8, evidenceCount BIGINT);;

CREATE TABLE association_by_datatype_indirect(diseaseId TEXT, targetId TEXT, datatypeId TEXT, score FLOAT8, evidenceCount BIGINT);;

CREATE TABLE association_by_overall_indirect(diseaseId TEXT, targetId TEXT, score FLOAT8, evidenceCount BIGINT);;

CREATE TABLE association_overall_direct(diseaseId TEXT, targetId TEXT, score FLOAT8, evidenceCount BIGINT);;

CREATE TABLE biosample(biosampleId TEXT, biosampleName TEXT, description TEXT, xrefs TEXT[], synonyms TEXT[], parents TEXT[], ancestors TEXT[], children TEXT[], descendants TEXT[]);;

CREATE TABLE colocalisation_coloc(leftStudyLocusId TEXT, rightStudyLocusId TEXT, chromosome TEXT, rightStudyType TEXT, numberColocalisingVariants BIGINT, h0 FLOAT8, h1 FLOAT8, h2 FLOAT8, h3 FLOAT8, h4 FLOAT8, colocalisationMethod TEXT, betaRatioSignAverage FLOAT8);;

CREATE TABLE colocalisation_ecaviar(leftStudyLocusId TEXT, rightStudyLocusId TEXT, chromosome TEXT, rightStudyType TEXT, numberColocalisingVariants BIGINT, clpp FLOAT8, colocalisationMethod TEXT, betaRatioSignAverage FLOAT8);;

CREATE TABLE credible_set(studyLocusId TEXT, studyId TEXT, variantId TEXT, chromosome TEXT, "position" INT, region TEXT, beta FLOAT8, zScore FLOAT8, pValueMantissa FLOAT, pValueExponent INT, effectAlleleFrequencyFromSource FLOAT, standardError FLOAT8, subStudyDescription TEXT, qualityControls TEXT[], finemappingMethod TEXT, credibleSetIndex INT, credibleSetlog10BF FLOAT8, purityMeanR2 FLOAT8, purityMinR2 FLOAT8, locusStart INT, locusEnd INT, sampleSize INT, ldSet STRUCT(tagVariantId TEXT, r2Overall FLOAT8)[], locus STRUCT(is95CredibleSet BOOL, is99CredibleSet BOOL, logBF FLOAT8, posteriorProbability FLOAT8, variantId TEXT, pValueMantissa FLOAT, pValueExponent INT, beta FLOAT8, standardError FLOAT8, r2Overall FLOAT8)[], confidence TEXT, studyType TEXT, isTransQtl BOOL);;

CREATE TABLE disease(id TEXT, code TEXT, "name" TEXT, description TEXT, dbXRefs TEXT[], parents TEXT[], synonyms STRUCT(hasExactSynonym TEXT[], hasRelatedSynonym TEXT[], hasNarrowSynonym TEXT[], hasBroadSynonym TEXT[]), obsoleteTerms TEXT[], obsoleteXRefs TEXT[], children TEXT[], ancestors TEXT[], therapeuticAreas TEXT[], descendants TEXT[], ontology STRUCT(isTherapeuticArea BOOL, leaf BOOL, sources STRUCT(url TEXT, "name" TEXT)));;

CREATE TABLE disease_hpo(id TEXT, "name" TEXT, description TEXT, dbXRefs TEXT[], parents TEXT[], obsoleteTerms TEXT[]);;

CREATE TABLE disease_phenotype(disease TEXT, phenotype TEXT, evidence STRUCT(aspect TEXT, bioCuration TEXT, diseaseFromSourceId TEXT, diseaseFromSource TEXT, diseaseName TEXT, evidenceType TEXT, frequency TEXT, modifiers TEXT[], onset TEXT[], qualifier TEXT, qualifierNot BOOL, "references" TEXT[], sex TEXT, resource TEXT)[]);;

CREATE TABLE drug_indication(id TEXT, indications STRUCT(disease TEXT, efoName TEXT, "references" STRUCT(source TEXT, ids TEXT[])[], maxPhaseForIndication FLOAT8)[], approvedIndications TEXT[], indicationCount INT);;

CREATE TABLE drug_mechanism_of_action(actionType TEXT, mechanismOfAction TEXT, chemblIds TEXT[], targetName TEXT, targetType TEXT, targets TEXT[], "references" STRUCT(source TEXT, ids TEXT[], urls TEXT[])[]);;

CREATE TABLE drug_molecule(id TEXT, canonicalSmiles TEXT, inchiKey TEXT, drugType TEXT, blackBoxWarning BOOL, "name" TEXT, yearOfFirstApproval BIGINT, maximumClinicalTrialPhase FLOAT8, parentId TEXT, hasBeenWithdrawn BOOL, isApproved BOOL, tradeNames TEXT[], synonyms TEXT[], crossReferences MAP(TEXT, TEXT[]), childChemblIds TEXT[], linkedDiseases STRUCT("rows" TEXT[], count INT), linkedTargets STRUCT("rows" TEXT[], count INT), description TEXT);;

CREATE TABLE drug_warning(chemblIds TEXT[], toxicityClass TEXT, country TEXT, description TEXT, id BIGINT, "references" STRUCT(ref_id TEXT, ref_type TEXT, ref_url TEXT)[], warningType TEXT, "year" BIGINT, efo_term TEXT, efo_id TEXT, efo_id_for_warning_class TEXT);;

CREATE TABLE evidence(datasourceId TEXT, targetId TEXT, alleleOrigins TEXT[], allelicRequirements TEXT[], ancestry TEXT, ancestryId TEXT, beta FLOAT8, betaConfidenceIntervalLower FLOAT8, betaConfidenceIntervalUpper FLOAT8, biologicalModelAllelicComposition TEXT, biologicalModelGeneticBackground TEXT, biologicalModelId TEXT, biomarkerName TEXT, biomarkers STRUCT(geneExpression STRUCT(id TEXT, "name" TEXT)[], geneticVariation STRUCT(functionalConsequenceId TEXT, id TEXT, "name" TEXT)[]), biosamplesFromSource TEXT[], cellType TEXT, clinicalPhase FLOAT8, clinicalSignificances TEXT[], clinicalStatus TEXT, cohortDescription TEXT, cohortId TEXT, cohortPhenotypes TEXT[], cohortShortName TEXT, confidence TEXT, contrast TEXT, crisprScreenLibrary TEXT, datatypeId TEXT, diseaseCellLines STRUCT(id TEXT, "name" TEXT, tissue TEXT, tissueId TEXT)[], diseaseFromSource TEXT, diseaseFromSourceId TEXT, diseaseFromSourceMappedId TEXT, diseaseModelAssociatedHumanPhenotypes STRUCT(id TEXT, "label" TEXT)[], diseaseModelAssociatedModelPhenotypes STRUCT(id TEXT, "label" TEXT)[], drugFromSource TEXT, drugId TEXT, drugResponse TEXT, geneticBackground TEXT, literature TEXT[], log2FoldChangePercentileRank BIGINT, log2FoldChangeValue FLOAT8, mutatedSamples STRUCT(functionalConsequenceId TEXT, numberMutatedSamples FLOAT8, numberSamplesTested FLOAT8, numberSamplesWithMutationType BIGINT)[], oddsRatio FLOAT8, oddsRatioConfidenceIntervalLower FLOAT8, oddsRatioConfidenceIntervalUpper FLOAT8, pValueExponent BIGINT, pValueMantissa FLOAT8, pathways STRUCT(id TEXT, "name" TEXT)[], projectId TEXT, reactionId TEXT, reactionName TEXT, releaseDate TEXT, releaseVersion TEXT, resourceScore FLOAT8, sex TEXT[], significantDriverMethods TEXT[], statisticalMethod TEXT, statisticalMethodOverview TEXT, statisticalTestTail TEXT, studyCases BIGINT, studyCasesWithQualifyingVariants BIGINT, studyId TEXT, studyOverview TEXT, studySampleSize BIGINT, studyStartDate TEXT, studyStopReason TEXT, studyStopReasonCategories TEXT[], targetFromSource TEXT, targetFromSourceId TEXT, targetInModel TEXT, targetInModelEnsemblId TEXT, targetInModelMgiId TEXT, targetModulation TEXT, urls STRUCT(niceName TEXT, url TEXT)[], variantAminoacidDescriptions TEXT[], variantFromSourceId TEXT, variantFunctionalConsequenceId TEXT, variantHgvsId TEXT, variantId TEXT, variantRsId TEXT, pmcIds TEXT[], publicationYear BIGINT, studyLocusId TEXT, textMiningSentences STRUCT(dEnd BIGINT, dStart BIGINT, section TEXT, tEnd BIGINT, tStart BIGINT, "text" TEXT)[], diseaseId TEXT, id TEXT, score FLOAT8, variantEffect TEXT, directionOnTrait TEXT, sourceId TEXT);;

CREATE TABLE expression(id TEXT, tissues STRUCT(efo_code TEXT, "label" TEXT, organs TEXT[], anatomical_systems TEXT[], rna STRUCT("value" FLOAT8, zscore INT, "level" INT, unit TEXT), protein STRUCT(reliability BOOL, "level" INT, cell_type STRUCT("name" TEXT, reliability BOOL, "level" INT)[]))[]);;

CREATE TABLE go(id TEXT, "name" TEXT);;

CREATE TABLE interaction(sourceDatabase TEXT, targetA TEXT, intA TEXT, intABiologicalRole TEXT, targetB TEXT, intB TEXT, intBBiologicalRole TEXT, speciesA STRUCT(mnemonic TEXT, scientific_name TEXT, taxon_id BIGINT), speciesB STRUCT(mnemonic TEXT, scientific_name TEXT, taxon_id BIGINT), count BIGINT, scoring FLOAT8);;

CREATE TABLE interaction_evidence(hostOrganismScientificName TEXT, hostOrganismTaxId BIGINT, expansionMethodShortName TEXT, targetB TEXT, evidenceScore FLOAT8, participantDetectionMethodB STRUCT(miIdentifier TEXT, shortName TEXT)[], intBBiologicalRole TEXT, interactionResources STRUCT(databaseVersion TEXT, sourceDatabase TEXT), interactionDetectionMethodShortName TEXT, intA TEXT, intBSource TEXT, interactionTypeMiIdentifier TEXT, expansionMethodMiIdentifier TEXT, speciesB STRUCT(mnemonic TEXT, scientificName TEXT, taxonId BIGINT), interactionIdentifier TEXT, speciesA STRUCT(mnemonic TEXT, scientificName TEXT, taxonId BIGINT), intASource TEXT, hostOrganismTissue STRUCT(fullName TEXT, shortName TEXT, xrefs TEXT[]), intB TEXT, pubmedId TEXT, intABiologicalRole TEXT, participantDetectionMethodA STRUCT(miIdentifier TEXT, shortName TEXT)[], interactionScore FLOAT8, interactionTypeShortName TEXT, targetA TEXT, interactionDetectionMethodMiIdentifier TEXT);;

CREATE TABLE known_drug(drugId TEXT, targetId TEXT, diseaseId TEXT, phase FLOAT8, status TEXT, urls STRUCT(niceName TEXT, url TEXT)[], ancestors TEXT[], "label" TEXT, approvedSymbol TEXT, approvedName TEXT, targetClass TEXT[], prefName TEXT, tradeNames TEXT[], synonyms TEXT[], drugType TEXT, mechanismOfAction TEXT, targetName TEXT);;

CREATE TABLE l2g_prediction(studyLocusId TEXT, geneId TEXT, score FLOAT8, features STRUCT("name" TEXT, "value" FLOAT, shapValue FLOAT)[], shapBaseValue FLOAT);;

CREATE TABLE literature(pmid TEXT, pmcid TEXT, date DATE, "year" INT, "month" INT, "day" INT, keywordId TEXT, relevance FLOAT8, keywordType TEXT);;

CREATE TABLE literature_vector(category TEXT, word TEXT, norm FLOAT8, vector FLOAT8[]);;

CREATE TABLE mouse_phenotype(biologicalModels STRUCT(allelicComposition TEXT, geneticBackground TEXT, id TEXT, literature TEXT[])[], modelPhenotypeClasses STRUCT(id TEXT, "label" TEXT)[], modelPhenotypeId TEXT, modelPhenotypeLabel TEXT, targetFromSourceId TEXT, targetInModel TEXT, targetInModelEnsemblId TEXT, targetInModelMgiId TEXT);;

CREATE TABLE openfda_significant_adverse_drug_reactions(chembl_id TEXT, "event" TEXT, count BIGINT, llr FLOAT8, critval FLOAT8, meddraCode TEXT);;

CREATE TABLE openfda_significant_adverse_target_reactions(targetId TEXT, "event" TEXT, count BIGINT, llr FLOAT8, critval FLOAT8, meddraCode TEXT);;

CREATE TABLE pharmacogenomics(datasourceId TEXT, datasourceVersion TEXT, datatypeId TEXT, directionality TEXT, evidenceLevel TEXT, genotype TEXT, genotypeAnnotationText TEXT, genotypeId TEXT, haplotypeFromSourceId TEXT, haplotypeId TEXT, literature TEXT[], pgxCategory TEXT, phenotypeFromSourceId TEXT, phenotypeText TEXT, studyId TEXT, targetFromSourceId TEXT, variantFunctionalConsequenceId TEXT, variantRsId TEXT, variantId TEXT, isDirectTarget BOOL, drugs STRUCT(drugFromSource TEXT, drugId TEXT)[]);;

CREATE TABLE reactome(id TEXT, "label" TEXT, ancestors TEXT[], descendants TEXT[], children TEXT[], parents TEXT[], path TEXT[][]);;

CREATE TABLE so(id TEXT, "label" TEXT);;

CREATE TABLE study(studyId TEXT, geneId TEXT, projectId TEXT, studyType TEXT, traitFromSource TEXT, traitFromSourceMappedIds TEXT[], biosampleFromSourceId TEXT, pubmedId TEXT, publicationTitle TEXT, publicationFirstAuthor TEXT, publicationDate TEXT, publicationJournal TEXT, backgroundTraitFromSourceMappedIds TEXT[], initialSampleSize TEXT, nCases INT, nControls INT, nSamples INT, cohorts TEXT[], ldPopulationStructure STRUCT(ldPopulation TEXT, relativeSampleSize FLOAT8)[], discoverySamples STRUCT(sampleSize INT, ancestry TEXT)[], replicationSamples STRUCT(sampleSize INT, ancestry TEXT)[], qualityControls TEXT[], analysisFlags TEXT[], summarystatsLocation TEXT, hasSumstats BOOL, condition TEXT, sumstatQCValues MAP(TEXT, FLOAT), diseaseIds TEXT[], backgroundDiseaseIds TEXT[], biosampleId TEXT);;

CREATE TABLE target(id TEXT, approvedSymbol TEXT, biotype TEXT, transcriptIds TEXT[], canonicalTranscript STRUCT(id TEXT, chromosome TEXT, "start" BIGINT, "end" BIGINT, strand TEXT), canonicalExons TEXT[], genomicLocation STRUCT(chromosome TEXT, "start" BIGINT, "end" BIGINT, strand INT), alternativeGenes TEXT[], approvedName TEXT, go STRUCT(id TEXT, source TEXT, evidence TEXT, aspect TEXT, geneProduct TEXT, ecoId TEXT)[], hallmarks STRUCT(attributes STRUCT(pmid BIGINT, description TEXT, attribute_name TEXT)[], cancerHallmarks STRUCT(pmid BIGINT, description TEXT, impact TEXT, "label" TEXT)[]), synonyms STRUCT("label" TEXT, source TEXT)[], symbolSynonyms STRUCT("label" TEXT, source TEXT)[], nameSynonyms STRUCT("label" TEXT, source TEXT)[], functionDescriptions TEXT[], subcellularLocations STRUCT("location" TEXT, source TEXT, termSL TEXT, labelSL TEXT)[], targetClass STRUCT(id BIGINT, "label" TEXT, "level" TEXT)[], obsoleteSymbols STRUCT("label" TEXT, source TEXT)[], obsoleteNames STRUCT("label" TEXT, source TEXT)[], "constraint" STRUCT(constraintType TEXT, score FLOAT, exp FLOAT, obs INT, oe FLOAT, oeLower FLOAT, oeUpper FLOAT, upperRank INT, upperBin INT, upperBin6 INT)[], tep STRUCT(targetFromSourceId TEXT, description TEXT, therapeuticArea TEXT, url TEXT), proteinIds STRUCT(id TEXT, source TEXT)[], dbXrefs STRUCT(id TEXT, source TEXT)[], chemicalProbes STRUCT(control TEXT, drugId TEXT, id TEXT, isHighQuality BOOL, mechanismOfAction TEXT[], origin TEXT[], probeMinerScore BIGINT, probesDrugsScore BIGINT, scoreInCells BIGINT, scoreInOrganisms BIGINT, targetFromSourceId TEXT, urls STRUCT(niceName TEXT, url TEXT)[])[], homologues STRUCT(speciesId TEXT, speciesName TEXT, homologyType TEXT, targetGeneId TEXT, isHighConfidence TEXT, targetGeneSymbol TEXT, queryPercentageIdentity FLOAT8, targetPercentageIdentity FLOAT8, priority INT)[], tractability STRUCT(modality TEXT, id TEXT, "value" BOOL)[], safetyLiabilities STRUCT("event" TEXT, eventId TEXT, effects STRUCT(direction TEXT, dosing TEXT)[], biosamples STRUCT(cellFormat TEXT, cellLabel TEXT, tissueId TEXT, tissueLabel TEXT)[], datasource TEXT, literature TEXT, url TEXT, studies STRUCT(description TEXT, "name" TEXT, "type" TEXT)[])[], pathways STRUCT(pathwayId TEXT, pathway TEXT, topLevelTerm TEXT)[], tss BIGINT);;

CREATE TABLE target_essentiality(id TEXT, geneEssentiality STRUCT(depMapEssentiality STRUCT(screens STRUCT(cellLineName TEXT, depmapId TEXT, diseaseCellLineId TEXT, diseaseFromSource TEXT, expression FLOAT8, geneEffect FLOAT8, mutation TEXT)[], tissueId TEXT, tissueName TEXT)[], isEssential BOOL)[]);;

CREATE TABLE target_prioritisation(targetId TEXT, isInMembrane INT, isSecreted INT, hasSafetyEvent INT, hasPocket INT, hasLigand INT, hasSmallMoleculeBinder INT, geneticConstraint FLOAT8, paralogMaxIdentityPercentage FLOAT8, mouseOrthologMaxIdentityPercentage FLOAT8, isCancerDriverGene INT, hasTEP INT, mouseKOScore FLOAT8, hasHighQualityChemicalProbes INT, maxClinicalTrialPhase FLOAT8, tissueSpecificity FLOAT8, tissueDistribution FLOAT8);;

CREATE TABLE variant(variantId TEXT, chromosome TEXT, "position" INT, referenceAllele TEXT, alternateAllele TEXT, variantEffect STRUCT("method" TEXT, assessment TEXT, score FLOAT, assessmentFlag TEXT, targetId TEXT, normalisedScore FLOAT8)[], mostSevereConsequenceId TEXT, transcriptConsequences STRUCT(variantFunctionalConsequenceIds TEXT[], aminoAcidChange TEXT, uniprotAccessions TEXT[], isEnsemblCanonical BOOL, codons TEXT, distanceFromFootprint BIGINT, distanceFromTss BIGINT, appris TEXT, maneSelect TEXT, targetId TEXT, impact TEXT, lofteePrediction TEXT, siftPrediction FLOAT, polyphenPrediction FLOAT, consequenceScore FLOAT, transcriptIndex INT, approvedSymbol TEXT, biotype TEXT, transcriptId TEXT)[], rsIds TEXT[], hgvsId TEXT, alleleFrequencies STRUCT(populationName TEXT, alleleFrequency FLOAT8)[], dbXrefs STRUCT(id TEXT, source TEXT)[], variantDescription TEXT);;

